src/hipscat_import/association/map_reduce.py

```diff
@@ -0,0 +1,204 @@
+            "Dir",
+            "Npix",
+            "join_Norder",
+            "join_Dir",
+            "join_Npix",
+            "primary_id",
+            "primary_hipscat_index",
+            "join_id",
+            "join_hipscat_index",
+        ]
+    ]
+
+    ## Write out association table shards.
+    join_data.to_parquet(
+        path=args.tmp_path,
+        engine="pyarrow",
+        partition_on=["Norder", "Dir", "Npix", "join_Norder", "join_Dir", "join_Npix"],
+        compute_kwargs={"partition_size": 1_000_000_000},
+        write_index=False,
+    )
+
+
+def reduce_association(input_path, output_path):
+    """Collate sharded parquet files into a single parquet file per partition"""
+    intermediate_partitions_file = file_io.append_paths_to_pointer(
+        input_path, "partitions.csv"
+    )
+    data_frame = file_io.load_csv_to_pandas(intermediate_partitions_file)
+
+    ## Clean up the dataframe and write out as our new partition join info file.
+    data_frame = data_frame[data_frame["primary_hipscat_index"] != 0]
+    data_frame["num_rows"] = data_frame["primary_hipscat_index"]
+    data_frame = data_frame[
+        ["Norder", "Dir", "Npix", "join_Norder", "join_Dir", "join_Npix", "num_rows"]
+    ]
+    data_frame = data_frame.sort_values(["Norder", "Npix", "join_Norder", "join_Npix"])
+    file_io.write_dataframe_to_csv(
+        dataframe=data_frame,
+        file_pointer=file_io.append_paths_to_pointer(
+            output_path, "partition_join_info.csv"
+        ),
+        index=False,
+    )
+
+    ## For each partition, join all parquet shards into single parquet file.
+    for _, partition in data_frame.iterrows():
+        input_dir = paths.create_hive_directory_name(
+            input_path,
+            ["Norder", "Dir", "Npix", "join_Norder", "join_Dir", "join_Npix"],
+            [
+                partition["Norder"],
+                partition["Dir"],
+                partition["Npix"],
+                partition["join_Norder"],
+                partition["join_Dir"],
+                partition["join_Npix"],
+            ],
+        )
+        output_dir = paths.pixel_association_directory(
+            output_path,
+            partition["Norder"],
+            partition["Npix"],
+            partition["join_Norder"],
+            partition["join_Npix"],
+        )
+        file_io.make_directory(output_dir, exist_ok=True)
+        output_file = paths.pixel_association_file(
+            output_path,
+            partition["Norder"],
+            partition["Npix"],
+            partition["join_Norder"],
+            partition["join_Npix"],
+        )
+        table = pq.read_table(input_dir)
+        rows_written = len(table)
+
+        if rows_written != partition["num_rows"]:
+            raise ValueError(
+                "Unexpected number of objects ",
+                f" Expected {partition['num_rows']}, wrote {rows_written}",
+            )
+
+        pq.write_table(table, where=output_file)
```

Comment:
What is the table here going to be sorted by, if anything? I'm not sure if it really matters for performance, but I think writing with pandas metadata here makes sense, and the best index column is probably the primary hipscat index